unit.cpp

```diff
@@ -535,17 +628,97 @@ void Unit::WriteReport(Areport *f, int obs, int truesight, int detfac,
 		temp += AString(WalkingCapacity());
 		temp += "/";
 		temp += AString(SwimmingCapacity());
+
 		temp += ". Skills: ";
 		temp += skills.Report(GetMen());
+		if (skills.Num())
+		{
+			of.StartArray("skills");
+			forlist(&skills)
+			{
+				Skill *s = (Skill*)elem;
+				of.StartDict(NULL);
+				of.PutPairStr("abbr", SkillDefs[s->type].abbr);
+				of.PutPairInt("level", GetLevelByDays(s->days/GetMen()));
+				of.PutPairInt("days", s->days/GetMen());
+				of.EndDict(); // end this skill
+			}
+			of.EndArray(); // end skills
+		}
 	}
 
 	if (obs == 2 && (type == U_MAGE || type == U_GUARDMAGE)) {
 		temp += MageReport();
+		if (combat != -1)
+			of.PutPairStr("combatSpell", SkillStrs(combat).Str());
 	}
 
 	if (obs == 2) {
 		temp += ReadyItem();
+		// JSON
+		bool first_weapon = true;
+		for (unsigned i = 0; i < MAX_READY; ++i)
+		{
+			const int ready = readyWeapon[i];
+			if (ready == -1)
+				continue;
+
+			if (first_weapon)
+			{
+				of.StartArray("readyWeapons");
+				first_weapon = false;
+			}
+			of.PutStr(ItemDefs[ready].abr);
+		}
+		if (!first_weapon)
+			of.EndArray();
+
+		bool first_armor = true;
+		for (unsigned i = 0; i < MAX_READY; ++i)
+		{
+			const int ready = readyArmor[i];
+			if (ready == -1)
+				continue;
+
+			if (first_armor)
+			{
+				of.StartArray("readyArmor");
+				first_armor = false;
+			}
+			of.PutStr(ItemDefs[ready].abr);
+		}
+		if (!first_armor)
+			of.EndArray();
+
+		// skills
 		temp += StudyableSkills();
+		of.StartArray("canStudy");
+		for (int sk = 0; sk < NSKILLS; ++sk)
+		{
+			// Magic skills without dependencies should not be skipped
+			if ((!SkillDefs[sk].depends[0].skill || SkillDefs[sk].depends[0].skill[0] == 0) && !(SkillDefs[sk].flags & SkillType::MAGIC))
+				continue;
+
+			// CanStudy() gets prereqs, does not check max level
+			if (!CanStudy(sk))
+				continue;
+
+			if (SkillDefs[sk].flags & SkillType::FOUNDATION)
+			{
+				// This is only used with mages to display studiable magic skills
+				if (type != U_MAGE)
+					continue;
+
+				// Compare current skill level to maximum level
+				// TODO find LCM man
```

Comment:
is there anything here needed to be finished?

Reply:
Only if you want to show "Can Study" for non-mages.